File: src/core/storage/keyValueStore.ts
import { createMMKV, type MMKV } from 'react-native-mmkv';

export interface KeyValueStore {
  read(key: string): string | null;
  write(key: string, value: string): void;
  remove(key: string): void;
}

export class InMemoryKeyValueStore implements KeyValueStore {
  protected readonly values = new Map<string, string>();

  read(key: string): string | null {
    return this.values.get(key) ?? null;
  }

  write(key: string, value: string): void {
    this.values.set(key, value);
  }

  remove(key: string): void {
    this.values.delete(key);
  }
}

export class ThrowingWriteStore extends InMemoryKeyValueStore {
  constructor(private readonly failingPrefix: string) {
    super();
  }

  write(key: string, value: string): void {
    if (key.startsWith(this.failingPrefix)) throw new Error('disk full');
    super.write(key, value);
  }
}

export class MmkvKeyValueStore implements KeyValueStore {
  private readonly mmkv: MMKV = createMMKV();

  read(key: string): string | null {
    return this.mmkv.getString(key) ?? null;
  }

  write(key: string, value: string): void {
    this.mmkv.set(key, value);
  }

  remove(key: string): void {
    this.mmkv.remove(key);
  }
}

File: lib/core/storage/key_value_store.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:shared_preferences/shared_preferences.dart';

abstract interface class KeyValueStore {
  String? read(String key);
  Future<void> write(String key, String value);
  Future<void> remove(String key);
}

class InMemoryKeyValueStore implements KeyValueStore {
  final Map<String, String> _values = {};

  @override
  String? read(String key) => _values[key];

  @override
  Future<void> write(String key, String value) async => _values[key] = value;

  @override
  Future<void> remove(String key) async => _values.remove(key);
}

class PrefsKeyValueStore implements KeyValueStore {
  PrefsKeyValueStore(this._prefs);

  final SharedPreferences _prefs;

  @override
  String? read(String key) => _prefs.getString(key);

  @override
  Future<void> write(String key, String value) => _prefs.setString(key, value);

  @override
  Future<void> remove(String key) => _prefs.remove(key);
}

final keyValueStoreProvider =
    Provider<KeyValueStore>((ref) => InMemoryKeyValueStore());
